vllm/model_executor/models/phi3v.py

```diff
@@ -303,231 +305,97 @@ def add_image_newline(self, image_features_hd):
-        image_data = [image_data]
-    elif is_list_of(image_data, torch.Tensor):
-        image_feature_size = [item.shape[0] for item in image_data]
-    else:
-        raise TypeError(f"Invalid image type: {type(image_data)}")
-
-    prompt = inputs.get("prompt")
-    if prompt is None:
-        # for async server request, we assume prompt and its token_ids is always
-        # in correct format. And num_image_tags == len(image_data) always True.
-        image_idx = range(1, len(image_data) + 1)
-        new_prompt = None
-    else:
-        image_idx = sorted(map(int, re.findall(r"<\|image_(\d+)\|>+", prompt)))
-        if prompt.count("<|image|>") > 0:
-            logger.warning("Please follow the prompt format that is "
-                           "documented on HuggingFace which does not involve "
-                           "repeating <|image|> tokens.")
-        elif (num_image_tags := len(image_idx)) > 1:
-            assert num_image_tags == len(
-                image_data), "The count of image_placeholder not match image's"
-        new_prompt = prompt
-
-    prompt_token_ids = inputs["prompt_token_ids"].copy()
-
-    # masked placeholder with image token id
-    for idx in image_idx:
-        candidates = _get_image_placeholder_token_id_candidates(model_config,
-                                                                idx=idx)
-
-        for candidate in candidates:
-            for i in range(len(prompt_token_ids) - len(candidate) + 1):
-                if prompt_token_ids[i:i + len(candidate)] == candidate:
-                    prompt_token_ids[i:i +
-                                     len(candidate)] = ([_IMAGE_TOKEN_ID] *
-                                                        len(candidate))
-                    break
-
-    # merge consecutive tag ids
-    merged_token_ids: List[int] = []
-    for is_placeholder, token_ids in itertools.groupby(
-            prompt_token_ids, lambda x: x == _IMAGE_TOKEN_ID):
-        if is_placeholder:
-            merged_token_ids.append(_IMAGE_TOKEN_ID)
-        else:
-            merged_token_ids.extend(list(token_ids))
-
-    # TODO: Move this to utils or integrate with clip.
-    new_token_ids: List[int] = []
-    placeholder_ranges: List[PlaceholderRange] = []
-    placeholder_idx = 0
-    while merged_token_ids:
-        token_id = merged_token_ids.pop(0)
-        if token_id == _IMAGE_TOKEN_ID:
-            replacement_ids = repeat_and_pad_token(
-                _IMAGE_TOKEN_ID,
-                repeat_count=image_feature_size[placeholder_idx],
-            )
-            placeholder_ranges.append({
-                "offset": len(new_token_ids),
-                "length": len(replacement_ids)
-            })
-            new_token_ids.extend(replacement_ids)
-            placeholder_idx += 1
-        else:
-            new_token_ids.append(token_id)
+    def _get_hf_processor(
+            self, mm_processor_kwargs: Mapping[str, object]) -> ProcessorMixin:
```

Comment:
Actually, I think it would be slightly more readable if we pass the dictionary as keyword arguments so we can explicitly state which kwargs are allowed.